extensions/migrate.md

```diff
@@ -0,0 +1,321 @@
+    message toward **OLD**, the message will eventually reach a server **X**
+    that has seen the "flip" message and be routed toward **NEW**. Since **X**
+    has already propagated the "flip", then the re-routed message will be seen
+    *after* the "flip" by each server along the path to **NEW**.
+
+  * If **NEW** will be buffering output as bytes, then **NEW** should have all
+    the information it needs to correctly format messages by the time it
+    receives the "flip" message. If this information is small, it could be
+    included in the "flip" itself. If this information is larger, or could
+    become larger, it should be sent before the "flip" as separate messages. If
+    this information can be changed externally, a smarter buffering technique
+    will be needed.
+
+  * Note that, after **OLD** has sent the "flip" message, but before the client
+    has resumed its session on **NEW**, changes to client state can only come
+    from other servers since **OLD** is no longer processing messages from the
+    client. This fact can make it easier to reason about the migration process.
+
+3. **OLD** sends any data it has about the client to **NEW** that **NEW** does
+   not already have, such as client capabilities. During this time, **NEW** is
+   receiving messages intended for the client. These messages should be
+   processed by **NEW** as if the client was connected, and any output should be
+   buffered.
+
+  * Note that **NEW** already has the vast majority of information it needs
+    about the client, such as nickname, user modes, services account name,
+    channel membership and status, away status, and so on. Very little data is
+    truly local only to **OLD**.
+
+  * None of the data transferred during this step or received from other servers
+    should affect buffering. If this is not the case, a smarter buffering
+    technique is necessary so that formatting can be delayed.
+
+4. Once all the data has has been transferred, **OLD** notifies the client with
+   `MIGRATE PROCEED` and can close the connection.
+
+5. After some time, the client will connect to **NEW** and request to resume the
+   session. After the client sends `MIGRATE END`, **NEW** drains its buffer.
+
+If any internal (non-leaf) servers do not understand the "flip" message, then
+migration is not possible. A leaf only needs to identify if a client is local or
+not, while internal servers take part in routing messages through the tree.
+Since leaves can become internal servers through regular routing rules, then all
+servers must be able to process the "flip" message for migration to be possible.
+
+# Protocol details
+
+## REQUEST
+
+Clients can send `MIGRATE REQUEST` to suggest a migration.
+
+    MIGRATE REQUEST cactus.desert :Free pizza
+
+The parameters are
+
+1. `REQUEST`, indicating the `MIGRATE REQUEST` operation
+
+2. The network-internal name of the server the client wishes to be migrated to.
```

Comment:
Where to find such name?

Reply:
This is the name in /map, /links, /whois, etc. The spec tries to distinguish between those names and DNS names, because there's not always a 1-1 mapping.